### PR TITLE
Add SPIDERMON_ENGINE_STOP_MONITORS setting to list monitors to be executed when Scrapy engine is stopped

### PR DESCRIPTION
**Pull request by HassanQamar07**
This implements the feature https://github.com/scrapinghub/spidermon/issues/82

### CODE REVIEW

**Comment by rennerocha** [comment]
As @ejulio said, code looks fine, but it would be very helpful if you could include some unit tests.

**Comment by raphapassini** [comment]
We are needing this feature in a project, I'll take care of the missing unit tests and documentation if you guys don't mind. @HassanQamar07 ?

**Comment by raphapassini** [comment]
I did add the test, there's a little problem but is due to the fact that `unittest.mock` used in `test_extension.py` is not available on Python2. We can easily fix this by adding the `mock` library to the `requirements.txt` but I would like to hear comments about the test first, maybe you guys have ideas about how to solve without mock, or you have more ideas of tests.

File: tests/test_extension.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+@pytest.fixture
+def crawler(spidermon_enabled_settings):
+    crawler = Crawler(Spider, settings=spidermon_enabled_settings)
+    crawler.spider = Spider('dummy')
+    return crawler
+
+
+def test_engine_stopped_suites_should_run(crawler):
+    "The suites defined at engine_stopped_suites should be loaded and run """
```
Comment (ejulio) [inline tests/test_extension.py]: Maybe I'm missing something, but there shouldn't be `""" some text """`

File: tests/test_extension.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    engine_stopped_suites = [
+        'tests.fixtures.suites.Suite01'
+    ]
+    spidermon = Spidermon(crawler, engine_stopped_suites=engine_stopped_suites)
+    spidermon.engine_stopped_suites[0].run = mock.MagicMock()
+    spidermon.engine_stopped()
+    assert spidermon.engine_stopped_suites[0].__class__.__name__ == 'Suite01'
+    assert spidermon.engine_stopped_suites[0].run.called
```
Comment (ejulio) [inline tests/test_extension.py]: I'm fine using `mock` and installing through `requirements`.
Here, probably I would assign `mock` to it's own variable to make the assertions shorter.
Also, probably would use `my_mock.assert_called_once()` or other assertion built-in with `mock`.

File: tests/test_extension.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+
+@pytest.fixture
+def crawler(spidermon_enabled_settings):
+    crawler = Crawler(Spider, settings=spidermon_enabled_settings)
+    crawler.spider = Spider('dummy')
+    return crawler
+
+
```
Comment (ejulio) [inline tests/test_extension.py]: Also, this test only tests that _when calling `engine_stopped`, it'll call the proper suites_.
I think that another important test would be to check if Spidermon subscribed properly to `engine_stopped` signal from Scrapy.

File: tests/test_extension.py (hunk 1 above)
Comment (rennerocha) [reply tests/test_extension.py]: I believe that when black is executed, this will be fixed.

**Comment by heylouiz** [comment]
@raphapassini @rennerocha @ejulio 
I made the requested changes, I had some problems to add `mock` as a requirement because I wanted to make sure it only installs when running with Python2, but I had problems with the tox configuration. I ended up putting it on a dedicated file for tests on requirements directory.

File: tests/test_extension.py (hunk 2 above)
Comment (heylouiz) [reply tests/test_extension.py]: @ejulio I tried to use `assert_called_once()` but it is not supported in python 3.5. I went back from Raphael's approach.

File: requirements/tests.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+mock
```
Comment (rennerocha) [inline requirements/tests.txt]: As this is required only for Python 2.7, you can include an environment marker, so it won't be installed in newer Python versions (https://www.python.org/dev/peps/pep-0508/#environment-markers)

File: spidermon/contrib/scrapy/extensions.py [hunk 5]
```diff
@@ -62,13 +67,16 @@ def from_crawler(cls, crawler):
...
+            engine_stopped_suites=crawler.settings.getlist('SPIDERMON_ENGINE_STOP_MONITORS'),
             spider_opened_expression_suites=crawler.settings.getlist('SPIDERMON_SPIDER_OPEN_EXPRESSION_MONITORS'),
             spider_closed_expression_suites=crawler.settings.getlist('SPIDERMON_SPIDER_CLOSE_EXPRESSION_MONITORS'),
+            engine_stopped_expression_suites=crawler.settings.getlist('SPIDERMON_ENGINE_STOP_EXPRESSION_MONITORS'),
             expressions_monitor_class=crawler.settings.get('SPIDERMON_EXPRESSIONS_MONITOR_CLASS'),
             periodic_suites=crawler.settings.getdict('SPIDERMON_PERIODIC_MONITORS'),
         )
         crawler.signals.connect(ext.spider_opened, signal=signals.spider_opened)
         crawler.signals.connect(ext.spider_closed, signal=signals.spider_closed)
+        crawler.signals.connect(ext.engine_stopped, signal=signals.engine_stopped)
```
Comment (rennerocha) [inline spidermon/contrib/scrapy/extensions.py]: This line could be tested like:
https://github.com/scrapinghub/spidermon/pull/120

File: tests/test_extension.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    return _crawler
+
+
+@pytest.fixture
+def suites():
+    return ['tests.fixtures.suites.Suite01']
+
+
+def test_spider_opened_suites_should_run(crawler, suites):
```
Comment (rennerocha) [inline tests/test_extension.py]: As `crawler` argument is a function, maybe a better name for the fixture would be `get_crawler`...

File: tests/test_extension.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    """The suites defined at spider_opened_suites should be loaded and run """
+    crawler = crawler()
```
Comment (rennerocha) [inline tests/test_extension.py]: ... then you could change to `crawler = get_crawler()` which looks more readable.

**Comment by rennerocha** [comment]
@heylouiz 
You need to apply black to format the code.

File: tests/test_extension.py (hunk 2 above)
Comment (ejulio) [reply tests/test_extension.py]: Are you sure?https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once

File: spidermon/contrib/scrapy/extensions.py (hunk 5 above)
Comment (ejulio) [reply spidermon/contrib/scrapy/extensions.py]: Nothing against the approach, I just think that we only need to test if it _connected_ to the signal.
The test of being called when the signal is triggered, is already handled by scrapy.

Comment (heylouiz) [reply spidermon/contrib/scrapy/extensions.py]: But I think it's best to test this on your PR  right? or on this if yours get approved first.

**Comment by heylouiz** [comment]
@ejulio 
https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once
`New in version 3.6.`

It was failing on python 3.5 build on Travis.

File: tests/test_extension.py (hunk 2 above)
Comment (heylouiz) [reply tests/test_extension.py]: @ejulio I replied on the wrong place, pasting here too:

https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once
`New in version 3.6.`

It was failing on python 3.5 build on Travis.

Comment (ejulio) [reply tests/test_extension.py]: My bad!
But, there are alternatives like `assert_any_call` or `assert_has_calls(1)`

Comment (heylouiz) [reply tests/test_extension.py]: @ejulio I ended up using assert_called_once_with(mock.ANY), what do you think?

Comment (ejulio) [reply tests/test_extension.py]: Looks good to me.
Not sure what is better, if `assert_any_call` or `assert_called_once_with(mock.ANY)`.

**Pull request merged**